src/integrator.jl

```diff
@@ -87,7 +87,13 @@ function step(
         else
             res = z
         end
-        !isfinite(z) && break
+        if !isfinite(z)
+            # Remove undef
+            if FullTraj
+                res = res[isassigned.(Ref(res), 1:n_steps)]
```

Comment:
@yebai @torfjelde I encountered an issue today. Basically the change in https://github.com/TuringLang/AdvancedHMC.jl/pull/200 reveals another uncatched bug before: the ummatrued chain (due to `break` caused by numerical issues) now could contain `undef`, which should be correctly "removed".

Reply:
Before we were passing a vector of `z` (initial phase point), which silently increase the probabilities of using the initial phase point if there is a numerical issue.

Reply:
Oooh, aight; nice catch!

Reply:
Is `Ref` necessary here, why? 

Also, is the broadcasting for vectorised chains?

Reply:
> Is Ref necessary here, why?

Yes, the signature is `isassigned(::Vector, ::Int)`.

> Also, is the broadcasting for vectorised chains?

What do you mean?

Reply:
Just to make the first reply easier to understand

```julia
julia> x = Vector{Any}(undef, 3)
3-element Array{Any,1}:
 #undef
 #undef
 #undef

julia> x[1] = 1
1

julia> x[2] = 2
2

julia> isassigned(x, 1)
true

julia> isassigned(x, 2)
true

julia> isassigned(x, 3)
false

julia> isassigned.(x, 1:3)
ERROR: MethodError: no method matching isassigned(::Int64, ::Int64)
Closest candidates are:
  isassigned(::Core.SimpleVector, ::Int64) at essentials.jl:651
  isassigned(::Array, ::Int64...) at array.jl:226
  isassigned(::BitArray, ::Int64) at bitarray.jl:107

julia> isassigned.(Ref(x), 1:3)
3-element BitArray{1}:
 1
 1
 0
```

Reply:
thanks for the clarification!